component/discovery/file/file.go

```diff
@@ -0,0 +1,346 @@
+package file
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/grafana/agent/component/discovery"
+
+	"github.com/bmatcuk/doublestar"
+	"github.com/fsnotify/fsnotify"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/component"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:    "discovery.file",
+		Args:    Arguments{},
+		Exports: Exports{},
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+// Arguments holds values which are used to configure the discovery.file
+// component.
+type Arguments struct {
+	Paths         []discovery.Target `river:"paths,attr"`
+	ExcludedPaths []discovery.Target `river:"excluded_paths,attr,optional"`
+	UpdatePeriod  time.Duration      `river:"update_period,attr,optional"`
+}
+
+func (a *Arguments) getPaths() []string {
+	paths := make([]string, 0)
+	index := 0
+	for _, v := range a.Paths {
+		val, found := v["__path__"]
+		if !found {
+			continue
+		}
+		paths = append(paths, val)
+		index++
+	}
+	return paths
+}
+
+func (a *Arguments) getExcluded() []string {
+	paths := make([]string, 0)
+	index := 0
+	for _, v := range a.Paths {
+		val, found := v["__path_exclude__"]
+		if !found {
+			continue
+		}
+		paths = append(paths, val)
+		index++
+	}
+	return paths
+}
+
+// Exports exposes targets.
+type Exports struct {
+	Targets []discovery.Target `river:"targets,attr"`
+}
+
+var _ component.Component = (*Component)(nil)
+
+// Component implements the discovery.file component.
+type Component struct {
+	opts component.Options
+
+	mut            sync.RWMutex
+	args           Arguments
+	watcher        *fsnotify.Watcher
+	watchesUpdated bool
+	watchedFiles   map[string]struct{}
```

Comment:
This has the same issue as the current implementation of `loki.source.file` where it can only keep track of each file _once_ and in this case, it's 'first one wins' (eg. it cannot have the same file or pattern with different labels twice).

If it's not practical to fix now, I think we should at least put out a TODO and/or document this behavior, but let's explore how we could improve this.